mgmt: mcumgr: Add optional image number to erase command

nordicjm opened the pull request:
Allows selecting which image will be erased, will default back to
image 1 as it would do in previous versions if the optional parameter
is not provided

Fixes #40855

nordicjm committed:
mgmt: mcumgr: Add optional slot number to erase command

Allows selecting which slot will be erased, will default back to
slot 1 as it would do in previous versions if the optional parameter
is not provided

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/zephyr_img_mgmt.c:
```diff
@@ -236,28 +247,43 @@ img_mgmt_vercmp(const struct image_version *a, const struct image_version *b)
 		}
 	}
 
+	flash_area_close(fa);
```
This close will be missed if line 274 fails.

nordicjm replied:
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/zephyr_img_mgmt.c:
```diff
@@ -80,24 +75,40 @@ zephyr_img_mgmt_flash_check_empty(uint8_t fa_id, bool *out_empty)
 }
 
+static int
+zephyr_img_mgmt_flash_check_empty(uint8_t fa_id, bool *out_empty)
```
Drop the zephyr_ prefix here, this is internal function and will not collide with anything else, and we do not support other systems so there is no point in making the function Zephyr specific.
We will weed out the zephyr_ prefix in other functions later.

nordicjm replied:
Fixed

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:
```diff
@@ -257,24 +257,59 @@ img_mgmt_erase(struct mgmt_ctxt *ctxt)
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Optional image number */
+	do {
+		struct zcbor_string key;
+		static const char image_keyword[] = "image";
+
+		ok = zcbor_tstr_decode(zsd, &key);
+
+		if (ok) {
+			if (key.len == (ARRAY_SIZE(image_keyword) - 1) &&
+			    memcmp(key.value, image_keyword, ARRAY_SIZE(image_keyword) - 1) == 0) {
+				break;
+			}
+
+			ok = zcbor_any_skip(zsd, NULL);
+		}
+	} while (ok);
+
+	if (ok) {
+		ok = zcbor_uint32_decode(zsd, &image);
+		if (!ok) {
+			return MGMT_ERR_EINVAL;
+		}
+	}
+
```
Can this be done with call to zcbor_decode_bulk with single decoding specification?

nordicjm replied:
Fixed

de-nordic reviewed (changes requested):
Some comments.
Is the commit with documentation change missing?

nordicjm committed:
doc: device_mgmt: smp: Add slot number to erase command

A slot number can now be provided to the mcumgr erase command to erase
a specific slot instead of slot 1, this documents the change in the
SMP group 1 documentation page.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release-notes: Add note on mcumgr img_mgmt erase slot

Adds details on the new optional slot number for the mcumgr img_mgmt
erase function.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
Release notes added and documentation updated

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:
```diff
@@ -257,24 +257,34 @@ img_mgmt_erase(struct mgmt_ctxt *ctxt)
+		ZCBOR_MAP_DECODE_KEY_VAL(slot, zcbor_uint32_decode, &slot),
+	};
+
+	ok = zcbor_map_decode_bulk(zsd, image_erase_decode,
```
`ok` is not used anywhere.
Should we even continue if this fails?

nordicjm replied:
Fixed

carlescufi commented:
@nordicjm please rebase
